File: rb/test/series_entity_test.rb
# Series entity test

require "minitest/autorun"
require "json"
require_relative "../FunisgoStreaming_sdk"
require_relative "runner"

class SeriesEntityTest < Minitest::Test
  def test_create_instance
    testsdk = FunisgoStreamingSDK.test(nil, nil)
    ent = testsdk.Series(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = series_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["create", "list", "update", "load", "remove"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "series." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set FUNISGOSTREAMING_TEST_SERIES_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # CREATE
    series_ref01_ent = client.Series(nil)
    series_ref01_data = Helpers.to_map(Vs.getprop(
      Vs.getpath(setup[:data], "new.series"), "series_ref01"))

    series_ref01_data_result, err = series_ref01_ent.create(series_ref01_data, nil)
    assert_nil err
    series_ref01_data = Helpers.to_map(series_ref01_data_result)
    assert !series_ref01_data.nil?
    assert !series_ref01_data["id"].nil?

    # LIST
    series_ref01_match = {}

    series_ref01_list_result, err = series_ref01_ent.list(series_ref01_match, nil)
    assert_nil err
    assert series_ref01_list_result.is_a?(Array)

    found_item = Vs.select(
      Runner.entity_list_to_data(series_ref01_list_result),
      { "id" => series_ref01_data["id"] })
    assert !Vs.isempty(found_item)

    # UPDATE
    series_ref01_data_up0_up = {
      "id" => series_ref01_data["id"],
    }

    series_ref01_markdef_up0_name = "created_at"
    series_ref01_markdef_up0_value = "Mark01-series_ref01_#{setup[:now]}"
    series_ref01_data_up0_up[series_ref01_markdef_up0_name] = series_ref01_markdef_up0_value

    series_ref01_resdata_up0_result, err = series_ref01_ent.update(series_ref01_data_up0_up, nil)
    assert_nil err
    series_ref01_resdata_up0 = Helpers.to_map(series_ref01_resdata_up0_result)
    assert !series_ref01_resdata_up0.nil?
    assert_equal series_ref01_resdata_up0["id"], series_ref01_data_up0_up["id"]
    assert_equal series_ref01_resdata_up0[series_ref01_markdef_up0_name], series_ref01_markdef_up0_value

    # LOAD
    series_ref01_match_dt0 = {
      "id" => series_ref01_data["id"],
    }
    series_ref01_data_dt0_loaded, err = series_ref01_ent.load(series_ref01_match_dt0, nil)
    assert_nil err
    series_ref01_data_dt0_load_result = Helpers.to_map(series_ref01_data_dt0_loaded)
    assert !series_ref01_data_dt0_load_result.nil?
    assert_equal series_ref01_data_dt0_load_result["id"], series_ref01_data["id"]

    # REMOVE
    series_ref01_match_rm0 = {
      "id" => series_ref01_data["id"],
    }
    _, err = series_ref01_ent.remove(series_ref01_match_rm0, nil)
    assert_nil err

    # LIST
    series_ref01_match_rt0 = {}

    series_ref01_list_rt0_result, err = series_ref01_ent.list(series_ref01_match_rt0, nil)
    assert_nil err
    assert series_ref01_list_rt0_result.is_a?(Array)

    not_found_item = Vs.select(
      Runner.entity_list_to_data(series_ref01_list_rt0_result),
      { "id" => series_ref01_data["id"] })
    assert Vs.isempty(not_found_item)

  end
end

def series_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "series", "SeriesTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = FunisgoStreamingSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["series01", "series02", "series03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["FUNISGOSTREAMING_TEST_SERIES_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "FUNISGOSTREAMING_TEST_SERIES_ENTID" => idmap,
    "FUNISGOSTREAMING_TEST_LIVE" => "FALSE",
    "FUNISGOSTREAMING_TEST_EXPLAIN" => "FALSE",
    "FUNISGOSTREAMING_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["FUNISGOSTREAMING_TEST_SERIES_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["FUNISGOSTREAMING_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["FUNISGOSTREAMING_APIKEY"],
      },
      extra || {},
    ])
    client = FunisgoStreamingSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["FUNISGOSTREAMING_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["FUNISGOSTREAMING_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end

File: go/test/series_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/funisgo-streaming-sdk/go"
	"github.com/voxgig-sdk/funisgo-streaming-sdk/go/core"

	vs "github.com/voxgig-sdk/funisgo-streaming-sdk/go/utility/struct"
)

func TestSeriesEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Series(nil)
		if ent == nil {
			t.Fatal("expected non-nil SeriesEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := seriesBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "series." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set FUNISGOSTREAMING_TEST_SERIES_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		seriesRef01Ent := client.Series(nil)
		seriesRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "series"}, setup.data), "series_ref01"))

		seriesRef01DataResult, err := seriesRef01Ent.Create(seriesRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		seriesRef01Data = core.ToMapAny(seriesRef01DataResult)
		if seriesRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if seriesRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		seriesRef01Match := map[string]any{}

		seriesRef01ListResult, err := seriesRef01Ent.List(seriesRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		seriesRef01List, seriesRef01ListOk := seriesRef01ListResult.([]any)
		if !seriesRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", seriesRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(seriesRef01List), map[string]any{"id": seriesRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		seriesRef01DataUp0Up := map[string]any{
			"id": seriesRef01Data["id"],
		}

		seriesRef01MarkdefUp0Name := "created_at"
		seriesRef01MarkdefUp0Value := fmt.Sprintf("Mark01-series_ref01_%d", setup.now)
		seriesRef01DataUp0Up[seriesRef01MarkdefUp0Name] = seriesRef01MarkdefUp0Value

		seriesRef01ResdataUp0Result, err := seriesRef01Ent.Update(seriesRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		seriesRef01ResdataUp0 := core.ToMapAny(seriesRef01ResdataUp0Result)
		if seriesRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if seriesRef01ResdataUp0["id"] != seriesRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if seriesRef01ResdataUp0[seriesRef01MarkdefUp0Name] != seriesRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", seriesRef01MarkdefUp0Name, seriesRef01ResdataUp0[seriesRef01MarkdefUp0Name])
		}

		// LOAD
		seriesRef01MatchDt0 := map[string]any{
			"id": seriesRef01Data["id"],
		}
		seriesRef01DataDt0Loaded, err := seriesRef01Ent.Load(seriesRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		seriesRef01DataDt0LoadResult := core.ToMapAny(seriesRef01DataDt0Loaded)
		if seriesRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if seriesRef01DataDt0LoadResult["id"] != seriesRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		seriesRef01MatchRm0 := map[string]any{
			"id": seriesRef01Data["id"],
		}
		_, err = seriesRef01Ent.Remove(seriesRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		seriesRef01MatchRt0 := map[string]any{}

		seriesRef01ListRt0Result, err := seriesRef01Ent.List(seriesRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		seriesRef01ListRt0, seriesRef01ListRt0Ok := seriesRef01ListRt0Result.([]any)
		if !seriesRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", seriesRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(seriesRef01ListRt0), map[string]any{"id": seriesRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func seriesBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "series", "SeriesTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read series test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse series test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"series01", "series02", "series03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("FUNISGOSTREAMING_TEST_SERIES_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"FUNISGOSTREAMING_TEST_SERIES_ENTID": idmap,
		"FUNISGOSTREAMING_TEST_LIVE":      "FALSE",
		"FUNISGOSTREAMING_TEST_EXPLAIN":   "FALSE",
		"FUNISGOSTREAMING_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["FUNISGOSTREAMING_TEST_SERIES_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["FUNISGOSTREAMING_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["FUNISGOSTREAMING_APIKEY"],
			},
			extra,
		})
		client = sdk.NewFunisgoStreamingSDK(core.ToMapAny(mergedOpts))
	}

	live := env["FUNISGOSTREAMING_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["FUNISGOSTREAMING_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
